### PR TITLE
refactor(expo-modules-autolinking): Replace `findModules` test suite

### PR DESCRIPTION
**Pull request by kitten**
# Why

- The `isNegativeTest` check was incorrect so wasn't actually run
- The directory structures were imprecise and the mocks were excessive
- The `projectRoot` wasn't always set correctly

Basically, the accuracy of the tests was questionable and it makes too many assumptions on the exact function calls and mocks, and the test validations were inaccurate, meaning, extending the tests in their current form to capture more expected behaviour with more edge cases wasn't feasible

Part of the problem this highlights is that `expo-modules-autolinking search` (and related commands) have too many brittle options imho, but that's unrelated to this PR.

# How

- Replace `require` calls in `findModules` implementation path with `fs.readFile` calls
  - **NOTE:** This isn't only for `memfs`, although that's a crucial change for it, but forcing ourselves to add wrapping calls here makes it easier to reason about (and extend) the loaders
- Replace `glob`/`require` mocks with `memfs`
- Rewrite test suite as nested `memfs` folder structures
- Fix `isNegativeTest` checks
- Replace all `expect()` checks with `toEqual` and `expect.objectContaining`
  - This is more accurate than length checks with `Object.keys(result).length` which is easy to omit and fool
  - This is more accurate than `toMatchObject`
- Add check for `realpath` mock to see how many calls it receives and capture that
- Add missing `mockRestore` default

# Test Plan

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

### CODE REVIEW

File: packages/expo-modules-autolinking/src/autolinking/__tests__/findModules-test.ts [hunk 1]
```diff
@@ -430,7 +424,69 @@ describe(findModulesAsync, () => {
...
+    });
+  });
+
+  /**
+   * TODO(@kitten): Fix this behaviour
+   * /app
+   *   ╚══ /node_modules/expo-module-1
+   *       └── /app/node_modules/expo-module-1/node_modules/expo-module-2
+   */
+  it('should link transitive non-hoisted module (FAILURE)', async () => {
```
Comment (kitten) [inline packages/expo-modules-autolinking/src/autolinking/__tests__/findModules-test.ts:485]: This, I expected to pass but doesn't. I know that some code basically resolves by dependencies by looking at `pkg.dependencies` (we'll likely have to extend this for `peerDependencies` since these can auto-install, and deduplicate target paths).

But that code/logic might be fighting the non-structural / non-traversing scanning logic. Something bails early and doesn't let this sub-dependency be discovered.

This is probably legacy behaviour that assumes that hoisting is always active

File: packages/expo-modules-autolinking/src/autolinking/__tests__/findModules-test.ts [hunk 2]
```diff
@@ -430,7 +424,69 @@ describe(findModulesAsync, () => {
...
+      '@expo/expo-test': expectAnyModule(),
+    });
+  });
+
+  /**
+   * TODO(@kitten): Documentation implies this should work
+   * /app
+   *   ├── /app/local-expo-module
+   */
+  it.skip('should link local non-node module (FAILURE)', async () => {
```
Comment (kitten) [inline packages/expo-modules-autolinking/src/autolinking/__tests__/findModules-test.ts]: This is behaviour that I think exists for `react-native-config` but for some reason doesn't seem to exist for Expo Modules? This probably will be a problem we have to reconcile when combining both algorithms (replacing `react-native-config` with `findModules`' algorithm)

I'm not sure why we allow `searchPaths` to ever change and be passed when `nativeModulesDir` is a thing? I'm also not sure if `nativeModulesDir` is always `node_modules`-structured (which I'd say is unexpected, since the name doesn't imply this) and whether it ignores nested `node_modules`.

We should probably remove `searchPaths` entirely imho

Comment (kitten) [reply packages/expo-modules-autolinking/src/autolinking/__tests__/findModules-test.ts]: I've found that this defaults to searching `./modules` from `projectRoot` (i.e. the app dir), which explains why it fails here. I updated the tests accordingly

File: packages/expo-modules-autolinking/src/platforms/__tests__/apple-test.ts [hunk 3]
```diff
@@ -215,29 +213,32 @@ describe(resolveModuleAsync, () => {
...
+    vol.fromJSON({ 'Podfile.properties.json': '' }, '/app/ios');
+
     const extraBuildDeps = await resolveExtraBuildDependenciesAsync('/app/ios');
     expect(extraBuildDeps).toBe(null);
   });
 
   it('should return null if Podfile.properties.json not found', async () => {
-    mockFsReadFile.mockRejectedValueOnce(new Error('File not found'));
+    vol.fromJSON({ 'Podfile.properties.json': null }, '/app/ios');
+
```
Comment (Kudo) [inline packages/expo-modules-autolinking/src/platforms/__tests__/apple-test.ts]:
```suggestion
```

maybe just don't add the file because the case is tested when the file not found

File: yarn.lock [hunk 4]
```diff
@@ -11111,6 +11180,16 @@ memfs@^3.2.0:
   dependencies:
     fs-monkey "^1.0.3"
 
+memfs@^4.17.2:
+  version "4.17.2"
+  resolved "https://registry.yarnpkg.com/memfs/-/memfs-4.17.2.tgz#1f71a6d85c8c53b4f1b388234ed981a690c7e227"
+  integrity sha512-NgYhCOWgovOXSzvYgUW0LQ7Qy72rWQMGGFJDoWg4G30RHd3z77VbYdtJ4fembJXBy8pMIUA31XNAupobOQlwdg==
+  dependencies:
+    "@jsonjoy.com/json-pack" "^1.0.3"
+    "@jsonjoy.com/util" "^1.3.0"
```
Comment (Kudo) [inline yarn.lock]: maybe align the memfs version as the other packages then to keep yarn.lock change smaller.

**Review by Kudo — APPROVED** [review]
